File: src/main/stack/python/Exec_Time.py
import time
import sys
import os
from statistics import median

from stack import Stack

class Exec_Time:
    """
    Classe responsável por medir e registrar os tempos de execução de operações em uma fila.
    Os resultados são salvos em arquivos separados para cada tipo de operação.
    """

    # Nomes dos métodos que serão testados e correspondem aos nomes dos arquivos de saída
    METHOD_NAMES = [
        "remove_last",
        "add_last",
        "get_last",
        "remove_first",
        "add_first",
        "get_first",
        "remove_middle",
        "add_middle",
        "get_middle"
    ]

    @staticmethod
    def main():
        """
        Método principal que coordena a execução do programa.
        Lê o arquivo de entrada, prepara os arquivos de saída e processa cada linha.
        """

        # Verifica se o número de argumentos é válido
        if len(sys.argv) < 2:
            print("ERRO: número de argumentos inválido!")
            return

        input_file = sys.argv[1] # Arquivo de entrada contendo os tamanhos das filas
        output_dir = "data/results/time/" # Diretório onde serão salvos os resultados

        try:
            # Cria o diretório de saída se não existir
            os.makedirs(output_dir, exist_ok=True)

            # Abre os arquivos de saída para escrita
            writers = []
            try:
                for method in ExecTime.METHOD_NAMES:
                    file_path = f"{output_dir}{method}.data"
                    # Abre o arquivo em modo de leitura e append
                    f = open(file_path, 'a+')
                    # Verifica se o arquivo está vazio e adiciona cabeçalho, se necessário
                    if os.fstat(f.fileno()).st_size == 0:
                        f.write("estrutura_linguagem tempo tamanho\n")
                    writers.append(f)
            except IOError as e:
                for f in writers:
                    f.close()
                raise

            # Conta o número total de linhas no arquivo para mostrar progresso
            with open(input_file, 'r') as reader:
                total_lines = sum(1 for line in reader if line.strip())

            # Processa cada linha do arquivo de entrada
            with open(input_file, 'r') as reader:
                current_line = 0
                for line in reader:
                    line = line.strip()
                    if not line:
                        continue

                    current_line += 1
                    print(f"Processando linha {current_line}/{total_lines}", end='\r')

                    # Converte a linha em uma lista de inteiros
                    elements = list(map(int, line.split()))
                    # Testa as operações com os elementos da linha atual
                    ExecTime.test_stack_operations(writers, elements)

            print(f"\rConcluído! Processadas {total_lines} linhas.")

        except IOError as e:
            print(f"Erro ao processar arquivos: {e}")
        finally:
            # Garante que todos os arquivos serão fechados
            for writer in writers:
                writer.close()

    @staticmethod
    def test_stack_operations(writers, elements):
        """
        Testa as operações da fila com os elementos fornecidos e registra os tempos.

        Args:
            writers: Lista de arquivos abertos para escrita dos resultados
            elements: Lista de elementos para preencher a fila inicialmente
        """
        length = len(elements) # Tamanho atual da fila
        middle = length // 2 # Índice do elemento do meio
        RUNS = 30 # Número de execuções para cada operação
        all_times = [[] for _ in range(len(ExecTime.METHOD_NAMES))] # Armazena tempos de cada operação

        # Preenche a fila com os elementos iniciais
        stack = stack(length)
        for num in elements:
            stack.add_last(num)

        # Executa os testes determinada quantia de vezes para extrair a mediana
        for run in range(RUNS):
            # Testa removeLast
            start_time = time.perf_counter_ns()
            stack.pop()
            end_time = time.perf_counter_ns()
            all_times[0].append(end_time - start_time)

            # Testa addLast
            start_time = time.perf_counter_ns()
            stack.push(999)
            end_time = time.perf_counter_ns()
            all_times[1].append(end_time - start_time)

            # Testa getLast
            start_time = time.perf_counter_ns()
            stack.peek()
            end_time = time.perf_counter_ns()
            all_times[2].append(end_time - start_time)

            # Testa removeFirst
            start_time = time.perf_counter_ns()
            stack.popLast()
            end_time = time.perf_counter_ns()
            all_times[3].append(end_time - start_time)

            # Testa addFirst
            start_time = time.perf_counter_ns()
            stack.pushLast(999)
            end_time = time.perf_counter_ns()
            all_times[4].append(end_time - start_time)

            # Testa getFirst
            start_time = time.perf_counter_ns()
            stack.peekLast()
            end_time = time.perf_counter_ns()
            all_times[5].append(end_time - start_time)

            # Testa remove (middle)
            start_time = time.perf_counter_ns()
            stack.popIndex(middle)
            end_time = time.perf_counter_ns()
            all_times[6].append(end_time - start_time)

            # Testa add (middle)
            start_time = time.perf_counter_ns()
            stack.pushIndex(999, middle)
            end_time = time.perf_counter_ns()
            all_times[7].append(end_time - start_time)

            # Testa get (middle)
            start_time = time.perf_counter_ns()
            stack.peekIndex(middle)
            end_time = time.perf_counter_ns()
            all_times[8].append(end_time - start_time)

        # Calcula a mediana dos tempos e escreve nos arquivos
        for i, method in enumerate(ExecTime.METHOD_NAMES):
            if all_times[i]:
                median_time = int(median(all_times[i]))
                writers[i].write(f"stack_python {median_time} {length}\n")
                writers[i].flush()

if __name__ == "__main__":
    ExecTime.main()

File: src/main/stack/cpp/Exec_time.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <chrono>
#include <stdexcept>
#include <filesystem>
#include "Stack.cpp"  

using namespace std;
using namespace std::chrono;
namespace fs = std::filesystem;

// Lista de nomes das operações que serão testadas
const vector<string> METHOD_NAMES = {
    "remove_last",
    "add_last",
    "get_last",
    "remove_first",
    "add_first",
    "get_first",
    "remove_middle",
    "add_middle",
    "get_middle"
};

/**
 * Testa todas as operações da Stack e registra os tempos de execução
 *
 * @param elements Vetor contendo os elementos iniciais da Stack
 * @param writers Array de arquivos de saída para gravar os resultados
 * @param currentLine Número da linha atual sendo processada (para exibir progresso)
 * @param totalLines Total de linhas a processar (para exibir progresso)
 */
void testStackOperations(const vector<int>& elements, ofstream writers[], int currentLine, int totalLines) {
    int length = elements.size();
    // Posição do meio para algumas operações
    int middle = length / 2;
    // Número de execuções para cada teste
    const int RUNS = 30;

    // Matriz para armazenar todos os tempos de execução
    vector<vector<long>> allTimes(METHOD_NAMES.size(), vector<long>(RUNS));

    // Cria e preenche a Stack com os elementos iniciais
    Stack stack;  // Supondo que a classe Stack tenha esse construtor
    for (int num : elements) {
        stack.addLast(num);  // Usando addLast para adicionar elementos
    }

    // Executa cada operação RUNS vezes para obter dados estatísticos
    for (int run = 0; run < RUNS; run++) {
        // Testa removeLast
        auto startTime = high_resolution_clock::now();
        stack.pop();
        auto endTime = high_resolution_clock::now();
        allTimes[0][run] = duration_cast<nanoseconds>(endTime - startTime).count();

        // Testa addLast
        startTime = high_resolution_clock::now();
        stack.push(999);  // Adiciona 999 no final
        endTime = high_resolution_clock::now();
        allTimes[1][run] = duration_cast<nanoseconds>(endTime - startTime).count();

        // Testa getLast
        startTime = high_resolution_clock::now();
        stack.peek();  // Obtém o último valor
        endTime = high_resolution_clock::now();
        allTimes[2][run] = duration_cast<nanoseconds>(endTime - startTime).count();

        // Testa removeFirst
        startTime = high_resolution_clock::now();
        stack.RemoveFirst();  // Remove o primeiro elemento
        endTime = high_resolution_clock::now();
        allTimes[3][run] = duration_cast<nanoseconds>(endTime - startTime).count();

        // Testa addFirst
        startTime = high_resolution_clock::now();
        stack.pushLast(999);  // Adiciona 999 no começo
        endTime = high_resolution_clock::now();
        allTimes[4][run] = duration_cast<nanoseconds>(endTime - startTime).count();

        // Testa getFirst
        startTime = high_resolution_clock::now();
        stack.peekLast();  // Obtém o primeiro valor
        endTime = high_resolution_clock::now();
        allTimes[5][run] = duration_cast<nanoseconds>(endTime - startTime).count();

        // Testa remove (middle)
        startTime = high_resolution_clock::now();
        stack.RemoveIndex(middle);  // Remove o valor no meio
        endTime = high_resolution_clock::now();
        allTimes[6][run] = duration_cast<nanoseconds>(endTime - startTime).count();

        // Testa add (middle)
        startTime = high_resolution_clock::now();
        stack.PushIndex(999, middle);  // Adiciona 999 no meio
        endTime = high_resolution_clock::now();
        allTimes[7][run] = duration_cast<nanoseconds>(endTime - startTime).count();
        
        // Testa get (middle)
        startTime = high_resolution_clock::now();
        stack.peekIndex(middle);  // Obtém o valor no meio
        endTime = high_resolution_clock::now();
        allTimes[8][run] = duration_cast<nanoseconds>(endTime - startTime).count();
    }

    // Processa os resultados: calcula a mediana para cada operação
    for (size_t method = 0; method < METHOD_NAMES.size(); method++) {
        sort(allTimes[method].begin(), allTimes[method].end());
        long medianTime = allTimes[method][RUNS / 2];
        
        writers[method] << "stack_cpp " << medianTime << " " << length << endl;
        writers[method].flush();
    }

    cerr << "\rProcessando linha " << currentLine << "/" << totalLines << flush;
}

/**
 * Função principal do programa
 *
 * @param argc Número de argumentos de linha de comando
 * @param argv Array de argumentos (argv[1] = arquivo de entrada)
 * @return 0 em caso de sucesso, 1 em caso de erro
 */
int main(int argc, char* argv[]) {
    if (argc < 2) {
        cerr << "ERRO: número de argumentos inválido!" << endl;
        return 1;
    }

    string inputFile = argv[1];
    string outputDir = "data/results/time/";

    try {
        fs::create_directories(outputDir);
    } catch (const fs::filesystem_error& e) {
        cerr << "ERRO: Não foi possível criar o diretório de saída: "
             << outputDir << "\n" << e.what() << endl;
        return 1;
    }

    ofstream writers[METHOD_NAMES.size()];
    for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
        string outputFile = outputDir + METHOD_NAMES[i] + ".data";
        bool fileExists = fs::exists(outputFile);
        bool fileIsEmpty = fileExists ? fs::file_size(outputFile) == 0 : true;
        writers[i].open(outputFile, ios::app);
        if (!writers[i]) {
            cerr << "\nErro ao abrir arquivo de saída: " << outputFile << endl;
            return 1;
        }
        if (!fileExists || fileIsEmpty) {
            writers[i] << "estrutura_linguagem tempo tamanho" << endl;
        }
    }

    try {
        ifstream reader(inputFile);
        if (!reader) {
            throw runtime_error("\nNão foi possível abrir o arquivo de entrada: " + inputFile);
        }

        int totalLines = 0;
        string line;
        while (getline(reader, line)) {
            if (!line.empty()) totalLines++;
        }
        reader.clear();
        reader.seekg(0);

        int currentLine = 0;
        while (getline(reader, line)) {
            if (line.empty()) continue;
            currentLine++;
            
            vector<int> elements;
            size_t pos = 0;
            while ((pos = line.find(' ')) != string::npos) {
                string token = line.substr(0, pos);
                elements.push_back(stoi(token));
                line.erase(0, pos + 1);
            }
            if (!line.empty()) {
                elements.push_back(stoi(line));
            }

            teststackOperations(elements, writers, currentLine, totalLines);
        }

        cerr << "\rConcluído! Processadas " << totalLines << " linhas." << endl;

        for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
            writers[i].close();
        }

    } catch (const exception& e) {
        cerr << "\nErro: " << e.what() << endl;
        for (size_t i = 0; i < METHOD_NAMES.size(); i++) {
            if (writers[i].is_open()) {
                writers[i].close();
            }
        }
        return 1;
    }

    return 0;
}
